### ISSUE TITLE
Look into refactoring code to be less reliant on added in byte flags

### ISSUE THREAD

**Issue by BlazinZzetti** [ShadowSpeedrun/ShadowSX#11]
Some features are currently relying on mod specific byte to check what the current game state is.

If possible, remove these byte flags to allow better compatibility with other game mods.

**Comment by BlazinZzetti** [comment]
Custom byte flags in SX 1.0 (Sorted by similar roles):

|Flag Name|Address|Byte Type|Saved Byte|Purpose|SX Only Feature|
|---|---|---|---|---|---|
|~~In Story Mode -~~ Button Check During Intro Messages|```0x8057D8F6```|Halfword|No|Flag to help handle features that should only occur in Story Mode. Is used initially for handing button inputs during the intro options messages. (Should be easily movable)|No/Yes|
|~~In Last Story~~|```0x8057D902```|Halfword|No|Flag to help handle features that should only occur in Last Story|No|
|~~Select Mode~~|```0x8057D8FE```|Halfword|No|Flag to help handle features that should only occur when playing in Select Mode|No|
|In Cutscene|```0x8057D8F8```|Halfword|No|Flag to help with managing Race Mode IGT|No|
|In Checkpoint|```0x8057D8A6```|Halfword|No|Handle various checks for keeping IGT running while in a checkpoint.|No|
|Egg Dealer Force Handicap|```0x8057D928```|Halfword|No|Allow Egg Dealer Handicap attack when Lives are above 0.|Yes|
|Allow Cutscene Skip Always|```0x80577B2C```|Byte|Yes|Flag used to allow skipping normally unskippable cutscenes|Yes|
|Race Mode|```0x80577B2D```|Byte|Yes|Option to display Total In Game Time instead of the Level Timer.|Yes|
|Modern UI Control|```0x80577B2E```|Byte|Yes|Swap X and B button actions to make UI control match Xbox and PS2|Yes|
|Show Race Time On Run Complete|```0x8057D900```|Halfword|No|Show Total In Game Time Message after game credits.|Yes|
|Intro Screen Index|```0x8057D8FA```|Halfword|No|Index value of which custom screen to show for SX Options and which code to run.|Yes|
|Intro Message ID|```0x8057D8FC```|Halfword|No|Index of message data to show for SX Options.|Yes|

**Comment by BlazinZzetti** [comment]
"SX Only Feature" meaning byte is needed for new feature and is not just for shortcutting checks for normal game states.  Flags with this option listed as "No" are candidates for removal if can be determine with mod agnostic code.

Intro Messages Bytes should be fine since no main game code should be running yet and can reasonably be overwritten once past those screens.

Saved bytes currently are stored in key data slots in unused stage save data, should be fine to leave untouched.

**Comment by BlazinZzetti** [comment]
#18 removes reliance on In Story Mode, In Last Story, and In Select Mode

**Comment by BlazinZzetti** [comment]
Pretty sure this is the updated list of custom ram values as of 7cae17f

Save Data
Flag Name | Address | Byte Size | Dual Use? | Purpose  
-- | -- | -- | -- | --
Story Mode Race Time Save | ```0x80577AF4``` | float | No | Storage of accumlated level times for this game mode.
Expert Race Time Save | ```0x80577B0C``` | float | No | Storage of accumlated level times for this game mode.
Last Story Race Time Save | ```0x80577B24``` | float | No | Storage of accumlated level times for this game mode.
Cutscene Skip | ```0x80577B2C``` | byte | No | Allow skipping a cutscene regardless of if it has been seen already.
Race Mode | ```0x80577B2D``` | byte | No | Show accumlated time instead of level time.
Modern UI Control | ```0x80577B2E``` | byte | No | Swap B and X functions for UI to match Xbox / PS2 menus.
Disable SPW Unlock | ```0x80577B2F``` | byte | No | Prevent SPW unlock on Story Mode ending.
Disable Key Saving (Not Practice Mode) | ```0x80577B30``` | byte | No | Prevent Keys Status Saving after level complete.

Intro Data
Flag Name | Address | Byte Size | Dual Use? | Purpose  
-- | -- | -- | -- | --
Intro Options Cursor Index | ```0x8057D8F4``` | halfword | Yes, Needs to ensure 0 on Intro Exit | Current selected options in Intro Options.
Intro Last Pressed Buttons   (Word) | ```0x8057D8F6``` | word | No | Last detected buttons used to only accept new inputs.
Custom FNT Offset | ```0x8057D8FA``` | halfword | No | When showing a expected message, this is use to show a custom one   instead.
Intro Message ID | ```0x8057D8FC``` | halfword | No | Used to detect if we have initialized the Intro Message.
Options Rendering Offsets | ```0x8057FB80``` | 10 halfwords | No | Byte offsets for where options and their settings should start rendering.
Page 2 Options Flags (Current Level Keys) | ```0x8057FBA0``` | 4 bytes | Yes / Not a concern | Temp storage of Page 2 Options which will execute upon Menu Exit
Available Options Enum | ```0x8057FBB0``` | 5 bytes | No | What the options are currently set to.
Options Mode Settings | ```0x8057FBB5``` | 5 bytes | No | What the options are allowed to be changed too.
Num of Options On Page | ```0x8057FBBA``` | byte | No | Number of options to process on the page.

Post Intro Data
Flag Name | Address | Byte Size | Dual Use? | Purpose  
-- | -- | -- | -- | --
"In Checkpoint" Flag | ```0x8057D8A6``` | byte | Yes, both not used at the same time. | Handle various checks for keeping IGT running while in a checkpoint.
"Prevent Time Stop and HUD Hide" Flag | ```0x8057D8A6``` | byte | Yes, both not used at the same time. | Allows showing the HUD and running timer during Egg Dealer Attacks.
Practice Mode | ```0x8057D8F4``` | byte | Yes, should be safe assuming clean up on Intro End | Used to tell custom code if extra options are available.
Practice Mode (Meter as Score Flag) | ```0x8057D8F5``` | byte | Yes, should be safe assuming clean up on Intro End | Used to tell UI code if the value used for scores should be meter points   instead.
Practice Mode (Input Lock) | ```0x8057D8F6``` | byte | Yes, should be safe assuming clean up on Intro End | Prevents PM codes from activating on input combo hold.
Practice Mode (Timer State) | ```0x8057D8F7``` | byte | Yes, should be safe assuming clean up on Intro End | Used to tell what the timer should be doing in PM.
"In Cutscene" Flag | ```0x8057D8F8``` | halfword | No | Flag to help with managing Race Mode IGT
Custom FNT Offset | ```0x8057D8FA``` | halfword | No | When showing a expected message, this is use to show a custom one   instead.
Show IGT at Auto Save | ```0x8057D900``` | halfword | Yes, should be managed correctly. | Set on Credits to show Race IGT before Auto Save
Boss Cutscene Flags (Intro) | ```0x8057D900``` | byte | Yes, cleaned up before Boss Fight End | Used to see if a boss intro cutscene has been seen for penalty timing.
Boss Cutscene Flags (Mid) | ```0x8057D901``` | byte | Yes, cleaned up before Boss Fight End | Used to see if a boss mid cutscene has been seen for penalty timing.
Temp Storage for Penalty Time | ```0x8057D904``` | float | No | Used as a parameter for a standalone function where time is added on to   the level time.
New IGT | ```0x8057D908``` | float | No | Level IGT, but with various logic applied to make it more like a loadless timer.
Race IGT | ```0x8057D918``` | float | No | All previous stages added together.
"Egg Dealer IL" Flag | ```0x8057D928``` | halfword | No | Allow Egg Dealer Handicap attack when Lives are above 0.

**Comment by BlazinZzetti** [comment]
So I believe this is a win for refactor but a fail for being less reliant. That said, I believe the overall goal of using OG game data for some game checks was accomplished. There's a lot that was added for SX1.1, especially with new Intro Options, Boss Penalties, and Practice Mode that I did pretty good considering the complexity of these features.  Besides a few fixes I found as I was making this list, besides maybe something for Turbo, I don't expect this to change much more if at all.  If so I will update the above comment.
